### ISSUE TITLE
Handle tests that should be run within the same process

### ISSUE THREAD

**Issue by sunshowers** [nextest-rs/nextest#27]
Currently, nextest runs every test in its own process. Should running tests within the same process be supported?

* [ ] How should test binaries be marked this way? `package.metadata` is one solution.
* [ ] How does concurrency control work? Rust's libtest doesn't support the jobserver protocol, so it's hard to communicate with it. One option is to run them serially at the end, once all the regular tests are run.
* [ ] How can test timing, stdout/stderr for individual tests, and other results be obtained? libtest's `--format json` is nightly-only. Maybe nextest can just give up on reporting individual statuses for these tests.

**Comment by wtfsck** [comment]
> Should running tests within the same process be supported?

Yes, assuming that's what `cargo test` does. If there are expensive `lazy_statics`, `cargo nextest` is slower than the normal `cargo test`.

**Comment by crepererum** [comment]
I think it would be nice if this is not an "all or nothing" flag, but rather that the user can describe multiple groups that should be run within a process, e.g. by specifying:

- workspace packages: each package in the list is a group
- multiple test-name wildcards: each wildcard is a group (not sure though how to guarantee that groups don't overlap)

All tests NOT found within a group are run within their own group.

To illustrate why this is helpful: over in [`influxdb_iox`](https://github.com/influxdata/influxdb_iox/) all tests within the workspace can be run in parallel. However for end2end tests within the `influxdb_iox` crate start a real server process but try to reuse it for multiple tests, so we pay for the startup costs every time. The tests using the single server process could also be run in parallel though, but currently they need to be executed from the same test process (they "find" and initialize the server process using `lazy_static`).

**Comment by sunshowers** [comment]
This is going to require somewhat advanced concurrency-related and stability-related work so is not a good first issue, but I'll mark this as "help wanted" for now. If you'd like to implement it, I'd be happy to accept a PR! We could probably go over further design intricacies once there's a PR for this.

**Comment by sunshowers** [comment]
Been thinking about this -- I've figured out a way to do this on stable Rust using the `--logfile` option. We can rely on two things:
1. `cargo test` executes tests in lexicographic order
2. data is written to logfiles as soon as the test passes

So what we can do is:
* execute the test binary with `--test-threads` set to `min(number of tests, cpu thread count)`
* instruct the client to write to a named pipe ([fifo](https://man7.org/linux/man-pages/man7/fifo.7.html) on unix, [named pipe](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea) on Windows)
* assume that tests are being executed in lexicographic order
* as results come in over the pipe, record test success/failure and time taken
* for `--success-output`/`--failure-output`, produce combined output from the binary at the end of execution

The time taken would likely be approximate, but that's ok -- we can document that.

**Comment by sunshowers** [comment]
Update: nextest now sets a [`NEXTEST_RUN_ID` environment variable](https://nexte.st/book/env-vars.html#environment-variables-nextest-sets) which should help porting over some use cases to nextest.

I don't plan to work on this for free. However:
* If someone is interested and willing to implement it, I would be happy to mentor them through it. I would estimate around 60-80 hours of work for someone new to the codebase to learn about it, implement this, write proper tests for it, work through edge cases and so on.
* If someone is interested and willing to pay me, I would be open to being contracted to work on this. Please reach out at inquiries@sunshowers.io to discuss this.

**Comment by sunshowers** [comment]
The ability to run tests with mutual exclusion (similar to the `serial_test` crate) is now possible with [test groups](https://nexte.st/book/test-groups.html).

**Comment by tisonkun** [comment]
Does nextest run concurrent cases in different process anyway?

I have a `OnceLock` logics and notice that it executes N times while it should not happen if tests are running in the same process ..

**Comment by sunshowers** [comment]
> Does nextest run concurrent cases in different process anyway?

Yes, the only supported model with nextest is process-per-test. See https://nexte.st/docs/configuration/test-groups/ for how to configure test groups.

**Comment by nazar-pc** [comment]
This issue is very important for running tests under Miri. On my 32C64T CPU running tests under Miri with and without nextest takes about the same amount of time, except nextest burns all CPU cores in the process and regular testing only a single core.

The reason seems to be the unusually high spawning cost of Miri, at least in this case, where it takes multiple seconds to even start a test. If only it could spawn the process once and run a bunch of tests there instead of exiting after one, then I'd be able to actually take advantage of nextest in a positive way.

This primarily applies to a large number of simple tests (I have unit tests for individual RISC-V instruction decoding and execution, hundreds of them) rather than small number of compute-intensive tests.

UPD: I came up with the following script to create batches and run tests in parallel, would be nice if nextest was capable of doing something similar:
```bash
#!/usr/bin/env bash
set -euo pipefail

tests=()
while IFS= read -r line; do
    tests+=("$line")
done < <(
    cargo miri test --lib --bins --tests -- --list --format terse 2>/dev/null \
        | grep ': test$' \
        | sed 's/: test$//'
)

cores=$(nproc 2>/dev/null || sysctl -n hw.logicalcpu)
total=${#tests[@]}
batch=$(( (total + cores - 1) / cores ))

pids=()

for (( i=0; i<total; i+=batch )); do
    (
        cargo miri test --lib --bins --tests -- --exact "${tests[@]:$i:$batch}" 2>&1 \
            | while IFS= read -r line; do
                echo "$line"
            done
    ) &
    pids+=($!)
done

for (( i=0; i<${#pids[@]}; i++ )); do
    if ! wait -n "${pids[@]}" 2>/dev/null; then
        kill "${pids[@]}" 2>/dev/null
        exit 1
    fi
done
```

**Comment by sunshowers** [comment]
Thanks. This was recently documented at https://nexte.st/docs/integrations/miri/#benefits.

It is unlikely nextest will support that kind of batch execution mode, for the reasons documented at https://nexte.st/docs/design/why-process-per-test/. I think the place to identify improvements is likely in Miri itself.

**Comment by sunshowers** [comment]
Put up a PR to use binary search while looking up tests with --exact, which should help your use case: https://github.com/rust-lang/rust/pull/154865

**Comment by sunshowers** [comment]
> Put up a PR to use binary search while looking up tests with --exact

This is somewhat faster now. With https://github.com/nextest-rs/many-tests/blob/main/tests/many.rs (1000 empty tests, basically a worst-case test for nextest) on a Ryzen 7950x (16c/32t), running 100 tests:

```
# Before:
$ rustc +nightly-2026-03-01 --version
rustc 1.96.0-nightly (38c0de8dc 2026-02-28)
$ time cargo +nightly-2026-03-01 miri nextest run -- test_1
cargo +nightly-2026-03-01 miri nextest run -- test_1  329.24s user 17.90s system 1876% cpu 18.498 total

# After:
$ rustc +nightly --version
rustc 1.97.0-nightly (9838411cb 2026-04-25)
$ time cargo +nightly miri nextest run -- test_1
cargo +nightly-2026-03-01 miri nextest run -- test_1  256.08s user 18.12s system 1727% cpu 15.876 total
```

That's around a 1.16x speedup. For more tests, the speedup will be greater since O(log N) vs O(N) becomes more prominent.

It is definitely still slower than regular cargo test, though:

```
$ time cargo +nightly miri test -- test_1
cargo +nightly miri test -- test_1  7.90s user 0.32s system 99% cpu 8.263 total
```

But as mentioned, this is a worst-case test: the test bodies are completely empty. The gap will become smaller with non-empty test bodies.

**Comment by RalfJung** [comment]
> This issue is very important for running tests under Miri.

The problem here is that Miri recompiles the test crate each time. We should find a way to not do that, which is tracked at https://github.com/rust-lang/miri/issues/5013.

> I think the place to identify improvements is likely in Miri itself.

Given the glaring bottleneck that currently exists here on the Miri side, agreed. Once that is resolved we should re-measure what things look like, maybe there's something that can be done by integrating Miri and nextest a bit more.

**Comment by RalfJung** [comment]
In the latest nightly, Miri got a lot better here -- we now use rustc's incremental cache to avoid redoing all the work of building the test crate. For big crates that still takes a lot longer than it should, but it's more than 2x faster than it was before.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/161868 should help some more to make `cargo miri nextest` faster. That's as much as I think we can currently do.

I'm not sure what the actionable part of this issue on the nextest side is. As far as the Miri part of it is concerned, I'd recommend closing it.

**Comment by sunshowers** [comment]
Thanks so much @RalfJung! Going to leave this issue open since it is more general, but glad to hear Miri is as good as it can be at this point.
